### DISCUSSION TITLE
setOpacity is not working on the Node

### DISCUSSION THREAD

**Discussion by HassamChaudhary1** [axmolengine/axmol#1484] [General]
setOpacity function on node is not working on my project. To check this, I also see the Demo project (Axmol test Project) and it is also not working on there.
setOpacity is working on the sprite but when I try to use it on the Node, it does not work.

Test Code :

```
ax::Node* testNode = Node::create();
    testNode->setPosition(100, 100);
    Sprite* testSprite = Sprite::create("LaunchScreenBackground.png");
    testNode->addChild(testSprite);
    testNode->setOpacity(100);
    addChild(testNode);
```

**Comment by HassamChaudhary1** [comment]
Anyone that can help me out bcoz I'm trying to investigate it since yesterday but I'm still on no output.

**Comment by dandy83ten** [comment]
testNode->setCascadeOpacityEnabled(true);

**Reply by HassamChaudhary1** [reply]
This will work for newly created node but it will not work for a full project. I checked on cocos-2dx (previous) and there Cascading is off and opacity is working there without adding any additional code.

**Reply by HassamChaudhary1** [reply]
I can't use setCascadeOpacityEnabled on my project bcoz I updated my project and previously it was working Fine and now why I need to do setCascadeOpacityEnabled?

**Reply by dandy83ten** [reply]
I know that setCascadeOpacityEnabled needs to be set.
_cascadeOpacityEnabled set default: false.

I don't know the details. I'm sorry I can't help youㅠㅠ

**Reply by HassamChaudhary1** [reply]
It's fine. Thanks for the response.

**Reply by dandy83ten** [reply]
cocos2d-x 3.x.x
![스크린샷 2023-12-08 오후 4 07 55](https://github.com/axmolengine/axmol/assets/26205032/a4faaf7c-19fa-47b4-b266-4e37317abe32)

axmol
![스크린샷 2023-12-08 오후 4 08 11](https://github.com/axmolengine/axmol/assets/26205032/89ccd242-ae23-4070-b03a-b4d424594079)

Do you use csb file?
The code in setPropsWithFlatBuffers has been changed.

Check it out.

**Reply by dandy83ten** [reply]
NodeReader::setPropsWithFlatBuffers

**Reply by HassamChaudhary1** [reply]
Yes I am using csb files. Let me take a look.

**Comment by rh101** [comment]
Since you're still using Cocos Studio, and not X-Studio, there is an easy way to fix this issue without having to modify the engine code.

Firstly though, please close the issue #1483 that you created, since this not a bug in the engine.

Now, to address your specific issue, try the following:

1 - Make a folder named "extensions" in the root of your project folder.
2 - Copy the `axmol/extensions/cocostudio` folder into this `[yourproject]/extensions` folder that you created.
3 - Disable the default Cocos Studio version that is in the engine source tree by adding this to your project `CMakeLists.txt`:
`set(AX_ENABLE_EXT_COCOSTUDIO OFF CACHE BOOL "Build extension cocostudio" FORCE)`

4 - Open `NodeReader.cpp` file in **your** project `extensions/cocostudio` folder, and edit the `WidgetReader::setPropsWithFlatBuffers` method to force the old behavior, which is to set both cascade color and opacity to `true`.

5 - Add this to your `CMakeLists.txt` to link with your modified version of that extension:
```
add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/extensions/cocostudio)
if(AX_ENABLE_EXT_SPINE)
  target_compile_definitions(cocostudio PRIVATE AX_ENABLE_EXT_SPINE=1)
  if(BUILD_SHARED_LIBS)
    target_link_libraries(cocostudio spine particle3d)
  else()
    get_target_property(spine_INCLUDE_DIRS spine INTERFACE_INCLUDE_DIRECTORIES)
    target_include_directories(cocostudio PUBLIC ${spine_INCLUDE_DIRS})
  endif()
endif()
if(AX_ENABLE_EXT_DRAGONBONES)
  target_compile_definitions(cocostudio PRIVATE AX_ENABLE_EXT_DRAGONBONES=1)
  if(BUILD_SHARED_LIBS)
    target_link_libraries(cocostudio DragonBones)
  endif()
endif()
```

That should work, and from that point you will be using your own version of the cocostudio extension, so you can modify it however you wish without having to change the engine source.

**Reply by HassamChaudhary1** [reply]
I'll close the bug. 
I don't get the point 4. Can you more specify point 4? @rh101

**Reply by rh101** [reply]
Say your project directory is named "MyGame", and you followed the above steps up to step 4. You would now have `MyGame/extensions/cocostudio`. So, in that directory tree, find the `NodeReader.cpp` file, and edit it with the changes you need.


Look at dandy83ten's post above to see what you need to change.

**Reply by HassamChaudhary1** [reply]
I have modified all the changes that you mentioned include CMake file but still opacity doesn't work. Check the attachment.

<img width="1080" alt="Screenshot 2023-12-08 at 4 21 52 PM" src="https://github.com/axmolengine/axmol/assets/79073135/f15b1ba8-077d-480d-96a5-68ee6b9217b9">

**Reply by rh101** [reply]
Did you delete the `CMakeCache.txt` file from the build folder? For example, if the build folder is `build_arm64`, then delete the file `build_arm64/CMakeCache.txt`, otherwise the cached `AX_ENABLE_EXT_COCOSTUDIO` may not be updated. Once you delete it, re-run the cmake command (or `axmol build ...`etc), whichever you originally used to generate the build folder, to recreate the project files.

Once you do recreate the project, rebuild your project, put a breakpoint on that code you changed, and make sure it is actually getting called.

**Reply by HassamChaudhary1** [reply]
Thank you @rh101 @dandy83ten for the help. Probably I missed something before.
With only these change, I can see the Opacity is working now. Thanks again.
<img width="1122" alt="Screenshot 2023-12-08 at 4 44 18 PM" src="https://github.com/axmolengine/axmol/assets/79073135/3072e2e9-a209-4ed6-b294-48e177bab125">
